workflow/controller/operator.go

```diff
@@ -534,17 +536,55 @@ func (woc *wfOperationCtx) requeue(afterDuration time.Duration) {
 }
 
 // processNodeRetries updates the retry node state based on the child node state and the retry strategy and returns the node.
-func (woc *wfOperationCtx) processNodeRetries(node *wfv1.NodeStatus, retryStrategy wfv1.RetryStrategy) (*wfv1.NodeStatus, error) {
+func (woc *wfOperationCtx) processNodeRetries(node *wfv1.NodeStatus, retryStrategy wfv1.RetryStrategy) (*wfv1.NodeStatus, bool, error) {
 	if node.Completed() {
-		return node, nil
+		return node, true, nil
 	}
 	lastChildNode, err := woc.getLastChildNode(node)
 	if err != nil {
-		return nil, fmt.Errorf("Failed to find last child of node " + node.Name)
+		return nil, false, fmt.Errorf("Failed to find last child of node " + node.Name)
 	}
 
 	if lastChildNode == nil {
-		return node, nil
+		return node, true, nil
+	}
+
+	if retryStrategy.Backoff != nil {
+		// Process max duration limit
+		if retryStrategy.Backoff.MaxDuration != "" && len(node.Children) > 0 {
+			maxDuration, err := parseStringToDuration(retryStrategy.Backoff.MaxDuration)
+			if err != nil {
+				return nil, false, err
+			}
+			firstChildNode, err := woc.getFirstChildNode(node)
+			if err != nil {
+				return nil, false, fmt.Errorf("Failed to find first child of node " + node.Name)
+			}
+			if time.Now().After(firstChildNode.FinishedAt.Add(maxDuration)) {
+				woc.log.Infoln("Max duration limit exceeded. Failing...")
+				return woc.markNodePhase(node.Name, lastChildNode.Phase, "Max duration limit exceeded"), true, nil
+			}
+		}
+
+		// Max duration limit hasn't been exceeded, process back off
+		if retryStrategy.Backoff.Duration == "" {
+			return nil, false, fmt.Errorf("Failed to find first child of node " + node.Name)
+		}
+		baseDuration, err := parseStringToDuration(retryStrategy.Backoff.Duration)
+		if err != nil {
+			return nil, false, err
+		}
+		// Formula: timeToWait = duration * factor^retry_number
+		timeToWait := baseDuration * time.Duration(math.Pow(float64(retryStrategy.Backoff.Factor), float64(len(node.Children))))
+		waitingDeadline := lastChildNode.FinishedAt.Add(timeToWait)
+
+		// See if we have waited past the deadline
+		if time.Now().Before(waitingDeadline) {
+			retryMessage := fmt.Sprintf("Retrying in %s", humanize.Duration(time.Until(waitingDeadline)))
+			return woc.markNodePhase(node.Name, node.Phase, retryMessage), false, nil
```

Comment:
A bit of an interesting trade-off here is that since we are changing the `parentNode` message, the workflow will be immediately re-queued. This makes re-queuing the Workflow with a duration (using `woc.requeue(time.Until(waitingDeadline)`) moot, and introduces some "polling" behavior in which the workflow is operated on once a second—mostly to update the "Retrying in ..." message.

If we decide that this polling behavior is not acceptable, then could make the "Retrying in ..." message static (e.g. "Retrying at [TIME]" and we could then re-queue with a duration as explained above.